### ISSUE TITLE
Compiled Query errors

### ISSUE THREAD

**Issue by soernt** [JasperFx/marten#2014]
Net 5.0
Marten 4.3.0

While running my test I get for each compiled query simular errors.
This is my query

```
using System;
using System.Linq;
using System.Linq.Expressions;
using Marten.AspNetCore.Identity.Models;
using Marten.Linq;

namespace Marten.AspNetCore.Identity.UserStore.Queries
{

    /// <summary>
    /// Determines if a user exists that as a given userId and roleId
    /// </summary>
    /// <typeparam name="TUser"></typeparam>
    public class UserFindByNormalizedUserNameQuery<TUser> : ICompiledQuery<TUser, TUser?>
        where TUser : MartenIdentityUser
    {
        private readonly string _normalizedUserName;

        /// <summary>
        /// Creates an instance of UserFindByNormalizedUserNameQuery
        /// </summary>
        /// <param name="normalizedUserName"></param>
        public UserFindByNormalizedUserNameQuery(string normalizedUserName)
        {
            _normalizedUserName = normalizedUserName;
        }

        /// <summary>
        /// Returns an expression that searches for a user by its normalized name value.
        /// </summary>
        /// <returns></returns>
        public Expression<Func<IMartenQueryable<TUser>, TUser?>> QueryIs()
        {
            return q => q.SingleOrDefault(x => x.NormalizedUserName == _normalizedUserName);
        }
    }
}
```

This is error I get:
```
System.InvalidOperationException: Compilation failures!

System.InvalidOperationException
Compilation failures!

CS1514: { erwartet.
CS1513: } erwartet.
CS1010: Zeilenvorschub in Konstante.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1010: Zeilenvorschub in Konstante.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS1514: { erwartet.
CS1513: } erwartet.
CS8803: Anweisungen der obersten Ebene müssen vor Namespace- und Typdeklarationen stehen.
CS1010: Zeilenvorschub in Konstante.
CS1002: ; erwartet.
CS1513: } erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS1010: Zeilenvorschub in Konstante.
CS1002: ; erwartet.
CS1526: Ein new-Ausdruck erfordert nach dem Typ eine Argumentliste oder (), [] oder {}.
CS1002: ; erwartet.
CS1010: Zeilenvorschub in Konstante.
CS1002: ; erwartet.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS1022: Typ- oder Namespacedefinition oder Dateiende erwartet.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS0116: Member, wie z. B. Felder oder Methoden, können nicht direkt in einem Namespace enthalten sein.
CS0115: "<invalid-global-code>.ConfigureCommand(CommandBuilder, IMartenSession)": Es wurde keine passende Methode zum Überschreiben gefunden.
CS0115: "<invalid-global-code>.BuildHandler(UserFindByNormalizedUserNameQuery<MartenIdentityUser>, IMartenSession)": Es wurde keine passende Methode zum Überschreiben gefunden.
CS8805: Das Programm mit Anweisungen der obersten Ebene muss eine ausführbare Datei sein.
CS0103: Der Name "_hardcoded" ist im aktuellen Kontext nicht vorhanden.
CS0103: Der Name "hardcoded" ist im aktuellen Kontext nicht vorhanden.
CS0103: Der Name "_maybeStatefulHandler" ist im aktuellen Kontext nicht vorhanden.
CS0103: Der Name "maybeStatefulHandler" ist im aktuellen Kontext nicht vorhanden.
CS0103: Der Name "_hardcoded" ist im aktuellen Kontext nicht vorhanden.

Code:

using Marten.AspNetCore.Identity.UserStore.Queries;
using Marten.Internal.CompiledQueries;
using Marten.Linq;
using Marten.Linq.QueryHandlers;
using System;

namespace Marten.Generated
{
    // START: UserFindByNormalizedUserNameQuery`1CompiledQuery
    public class UserFindByNormalizedUserNameQuery"1CompiledQuery : Marten.Internal.CompiledQueries.ClonedCompiledQuery<Marten.AspNetCore.Identity.Models.MartenIdentityUser, Marten.AspNetCore.Identity.UserStore.Queries.UserFindByNormalizedUserNameQuery<Marten.AspNetCore.Identity.Models.MartenIdentityUser>>
    {
        private readonly Marten.Linq.QueryHandlers.IMaybeStatefulHandler _inner;
        private readonly Marten.AspNetCore.Identity.UserStore.Queries.UserFindByNormalizedUserNameQuery<Marten.AspNetCore.Identity.Models.MartenIdentityUser> _query;
        private readonly Marten.Linq.QueryStatistics _statistics;
        private readonly Marten.Internal.CompiledQueries.HardCodedParameters _hardcoded;

        public UserFindByNormalizedUserNameQuery"1CompiledQuery(Marten.Linq.QueryHandlers.IMaybeStatefulHandler inner, Marten.AspNetCore.Identity.UserStore.Queries.UserFindByNormalizedUserNameQuery<Marten.AspNetCore.Identity.Models.MartenIdentityUser> query, Marten.Linq.QueryStatistics statistics, Marten.Internal.CompiledQueries.HardCodedParameters hardcoded) : base(inner, query, statistics, hardcoded)
        {
            _inner = inner;
            _query = query;
            _statistics = statistics;
            _hardcoded = hardcoded;
        }



        public override void ConfigureCommand(Weasel.Postgresql.CommandBuilder builder, Marten.Internal.IMartenSession session)
        {
            var parameters = builder.AppendWithParameters(@"select d.id, d.data, d.mt_version from integrationtest_136a328ca12d421c9aee1b0ecb7826c5.mt_doc_martenidentityuser as d where d.data ->> 'NormalizedUserName' = ? LIMIT ?");
            _hardcoded.Apply(parameters);
        }

    }

    // END: UserFindByNormalizedUserNameQuery`1CompiledQuery
    
    
    // START: UserFindByNormalizedUserNameQuery`1CompiledQuerySource
    public class UserFindByNormalizedUserNameQuery"1CompiledQuerySource : Marten.Internal.CompiledQueries.CompiledQuerySource<Marten.AspNetCore.Identity.Models.MartenIdentityUser, Marten.AspNetCore.Identity.UserStore.Queries.UserFindByNormalizedUserNameQuery<Marten.AspNetCore.Identity.Models.MartenIdentityUser>>
    {
        private readonly Marten.Internal.CompiledQueries.HardCodedParameters _hardcoded;
        private readonly Marten.Linq.QueryHandlers.IMaybeStatefulHandler _maybeStatefulHandler;

        public UserFindByNormalizedUserNameQuery"1CompiledQuerySource(Marten.Internal.CompiledQueries.HardCodedParameters hardcoded, Marten.Linq.QueryHandlers.IMaybeStatefulHandler maybeStatefulHandler)
        {
            _hardcoded = hardcoded;
            _maybeStatefulHandler = maybeStatefulHandler;
        }



        public override Marten.Linq.QueryHandlers.IQueryHandler<Marten.AspNetCore.Identity.Models.MartenIdentityUser> BuildHandler(Marten.AspNetCore.Identity.UserStore.Queries.UserFindByNormalizedUserNameQuery<Marten.AspNetCore.Identity.Models.MartenIdentityUser> query, Marten.Internal.IMartenSession session)
        {
            return new Marten.Generated.UserFindByNormalizedUserNameQuery"1CompiledQuery(_maybeStatefulHandler, query, null, _hardcoded);
        }

    }

    // END: UserFindByNormalizedUserNameQuery`1CompiledQuerySource
    
    
}


   at LamarCompiler.AssemblyGenerator.Generate(String code)
   at LamarCompiler.AssemblyGenerator.Compile(GeneratedAssembly generatedAssembly, IServiceVariableSource services)
   at Marten.Internal.CompiledQueries.CompiledQuerySourceBuilder.Build()
   at Marten.StoreOptions.GetCompiledQuerySourceFor[TDoc,TOut](ICompiledQuery`2 query, IMartenSession session)
   at Marten.Internal.Sessions.QuerySession.QueryAsync[TDoc,TOut](ICompiledQuery`2 query, CancellationToken token)
   at Marten.AspNetCore.Identity.UserStore.MartenUserStore`1.FindByNameAsync(String normalizedUserName, CancellationToken cancellationToken) in C:\Projekte\CSharp\AspNetCore.Identity.Marten\source\Marten.AspNetCore.Identity\UserStore\MartenUserStore.cs:line 331
   at Microsoft.AspNetCore.Identity.UserManager`1.FindByNameAsync(String userName)
   at Microsoft.AspNetCore.Identity.UserValidator`1.ValidateUserName(UserManager`1 manager, TUser user, ICollection`1 errors)
   at Microsoft.AspNetCore.Identity.UserValidator`1.ValidateAsync(UserManager`1 manager, TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.ValidateUserAsync(TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.CreateAsync(TUser user)
   at Microsoft.AspNetCore.Identity.Test.IdentitySpecificationTestBase`3.AddUserToRoleFailsIfAlreadyInRole()
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

**Comment by jeremydmiller** [comment]
@soernt Using an open generic type for a compiled query isn't supported. That's all you've bumped into here. Can't say that I ever anticipated that. If you make a closed type subclass of that for each `TUser` type, that *should* work

**Comment by jeremydmiller** [comment]
Or we could try to patch the code generation to handle this. Might not be that big of an issue

**Comment by jeremydmiller** [comment]
And I've got it fixed, will be in 4.3.1.

**Comment by soernt** [comment]
:-) Thank you
